pkg/operatorstatus/operatorstatus.go

```diff
@@ -59,44 +58,63 @@ type ClusterOperatorWrapper interface {
 }
 
 func (o *ClusterOperatorHandler) UpdateOperatorStatus(sampleResource *v1alpha1.SamplesResource) error {
+
 	var err error
+
 	switch {
-	case sampleResource.ConditionTrue(v1alpha1.SamplesExist):
-		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionTrue, "Samples exist in the openshift project")
+	// samples exists == true means samples created; versions matching means imagestream imports done
+	case sampleResource.ConditionTrue(v1alpha1.SamplesExist) && sampleResource.Spec.Version == sampleResource.Status.Version:
+		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionTrue, fmt.Sprintf("Samples installed at %s", sampleResource.Status.Version), "")
 	case sampleResource.ConditionFalse(v1alpha1.SamplesExist):
-		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionFalse, "Samples do not exist in the openshift project")
-	default:
-		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionUnknown, "The presence of the samples in the openshift project is unknown")
+	case sampleResource.ConditionUnknown(v1alpha1.SamplesExist):
+		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionFalse, fmt.Sprintf("Samples not yet installed at %s", sampleResource.Spec.Version), "")
 	}
 	if err != nil {
 		return err
 	}
 
+	errorDetails := sampleResource.ConditionsMessages()
 	switch {
-	case sampleResource.ConditionTrue(v1alpha1.ConfigurationValid):
-		err = o.setOperatorStatus(configv1.OperatorFailing, configv1.ConditionFalse, "The samples operator configuration is valid")
+	// errorDetails should have data if config valid false or progress unknown, but just in case
 	case sampleResource.ConditionFalse(v1alpha1.ConfigurationValid):
-		err = o.setOperatorStatus(configv1.OperatorFailing, configv1.ConditionTrue, "The samples operator configuration is invalid")
+	case sampleResource.ConditionUnknown(v1alpha1.ImageChangesInProgress):
+	case len(errorDetails) > 0:
+		err = o.setOperatorStatus(configv1.OperatorFailing, configv1.ConditionTrue, fmt.Sprintf("Unable to install samples at %s: %s", sampleResource.Spec.Version, errorDetails), "")
+		if err != nil {
+			return err
+		}
+		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionFalse, fmt.Sprintf("Unable to install samples at %s", sampleResource.Spec.Version), "")
```

Comment:
is available really guaranteed to be false here?  

If configurationvalid==false but we previously had a valid configuration, then we would have (may have?) already installed the samples so things are "Available" even if we can't progress to the new desired state.  Basically if "samplesexist==true" && "imagechangesinprogress==false" then i think available should be true?


if imagechangesinprogress==true then i guess i agree, it's not available.

Reply:
You are right on an upgrade scenario it could still be true at the older level.  I'll make a note of covering this with the next iteration choice we make below.

Reply:
it's not just on the upgrade scenario.  I could also start out in a valid config, get everything created, and then change the config to something invalid, right?

Reply:
(at which point the samples created by the original valid config would still exist, so we'd be in available=true, progressing=false, failing=true)

Reply:
yes I did not mean to imply the upgrade scenario was the exhaustive list